### PR TITLE
updated interface for accumulation

### LINKED ISSUE #547: Distinguish NA (missing) from NA (accumulated)
> Enabling this would, I think, require some sort of that marks dates explicitly as missing vs. NA.

I think this would be my preferred option as it would be more general but I also think it can be addressed in its own review as it would be a superset of this PR. 

My thought on how that would work is to have a new variable (`accumulate`) that indicates which days should be summed.

_Originally posted by @seabbs in https://github.com/epiforecasts/EpiNow2/pull/534#pullrequestreview-1880071190_

### PR DESCRIPTION
**Pull request by sbfnk**
## Description

This PR closes #547 following the design agreed in #839.

## Initial submission checklist

- [X] My PR is based on a package issue and I have explicitly linked it.
- [X] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [X] I have added or updated unit tests where necessary.
- [X] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [X] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [X] I have added a news item linked to this PR.

## After the initial Pull Request 

- [x] I have reviewed Checks for this PR and addressed any issues as far as I am able.

### CODE REVIEW

File: R/preprocessing.R [hunk 1]
```diff
@@ -0,0 +1,152 @@
+##' Fills missing data in a data set to prepare it for use within the package
+##'
+##' @description `r lifecycle::badge("experimental")`
+##'
+##' @param data Data frame with a `date` column. The other columns depend on the
+##'   model that the data are to be used, e.g. [estimate_infections()] or
+##'   [estimate_secondary()]. See the documentation there for the expected
+##'   format.
+##' @param missing Character. Options are "missing" (the default), "accumulate"
+##'   and "zero". This determines how missing dates the data are interpreted.
```
Comment (seabbs) [inline R/preprocessing.R:10]:
```suggestion
##'   and "zero". This determines how missing dates in the data are interpreted.
```

File: R/preprocessing.R [hunk 2]
```diff
@@ -0,0 +1,152 @@
+##' Fills missing data in a data set to prepare it for use within the package
+##'
+##' @description `r lifecycle::badge("experimental")`
+##'
+##' @param data Data frame with a `date` column. The other columns depend on the
+##'   model that the data are to be used, e.g. [estimate_infections()] or
+##'   [estimate_secondary()]. See the documentation there for the expected
+##'   format.
+##' @param missing Character. Options are "missing" (the default), "accumulate"
```
Comment (seabbs) [inline R/preprocessing.R:9]: When I first read this I didn't totally get what it was for but after a few passes I did. I wondered if a change name or a little sentence more clearly stating what missing was here would help but I couldn't come up with anything so perhaps it is just a difficult thing that people need to take the time to pass

Comment (seabbs) [reply R/preprocessing.R:9]: I think the thing that confused me most was `missing = "missing"`. Perhaps it should be `missing = "na"?

File: R/create.R [hunk 3]
```diff
@@ -525,14 +528,16 @@ create_obs_model <- function(obs = obs_opts(), dates) {
...
-
   cases <- data[(seeding_time + 1):(.N - horizon)]
   complete_cases <- create_complete_cases(cases)
   cases <- cases$confirm
+  accumulate <- data[-(1:seeding_time)]$accumulate
 
   stan_data <- list(
     cases = complete_cases$confirm,
     cases_time = complete_cases$lookup,
+    any_accumulate = as.integer(any(accumulate > 0)),
```
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/create.R:539]: Nope

```r
> any(isTRUE(c(TRUE, FALSE, TRUE)))
[1] FALSE
```

File: R/preprocessing.R [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+##' @param na Character. How to process dates that have NA values. The options
+##'   available are the same ones as for the `missing` argument.
+##' @param initial_accumulate Integer. The number of initial dates to accumulate
+##'   if `missing` or a column name is set to `"accumulate"`. This number of
+##'   dates is added to the beginning of the data set to be accumulated onto the
+##'   first data point. This is useful, for example, for modelling weekly
+##'   incidence data, in which case this should be set to 7. If accumulating and
+##'   the first data point is one to accumulate and this is not set, then that
+##'   data point will be removed with a warning.
+##' @param obs_column Character. If given, only the column specified here will
```
Comment (seabbs) [inline R/preprocessing.R:29]: This auto-magic feature seems like it could be dangerous as people might not expect random columns epinow2 doesn't use to do something like this. Is there a reason to not make the default the core set (i.e. `confirm`)?

File: R/preprocessing.R [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+##'   be used for checking missingness. This is useful if using a data set that
+##'   has NA values in multiple columns, but only one of them corresponds to
+##'   observations that are to be processed here. By default, all columns will
+##'   be processed and, e.g. `accumulate` will be set if any column is NA in any
+##'   given row.
+##' @return a data.table with an `accumulate` column that indicates whether
+##'   values are accumulated (see the documnetation of the `data` arugment in
```
Comment (seabbs) [inline R/preprocessing.R:36]:
```suggestion
##'   values are accumulated (see the documentation of the `data` argument in
```

File: R/preprocessing.R [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+  date_seq <- seq.Date(
+    min(data$date) - initial_add + 1,
+    max(data$date), by = "day"
+  )
+
+  ## mark dates that are present in the data
+  data[, ..present := TRUE]
+  complete_dates <- data.table(date = date_seq)
+  data <- merge(data, complete_dates, by = "date", all.y = TRUE)
```
Comment (seabbs) [inline R/preprocessing.R:81]: if people are preparing data from multi-regions will this break? does it matter if it does

File: R/create.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/create.R:539]: Alternatively, we could do `as.integer(TRUE %in% accumulate)`.

File: R/preprocessing.R (hunk 4 above)
Comment (seabbs) [reply R/preprocessing.R:29]: is the expectation that users will nearly always set this themselves? If yes maybe we throw a message if they don't set it?

File: inst/stan/estimate_infections.stan [hunk 7]
```diff
@@ -139,6 +139,13 @@ transformed parameters {
     }
   }
 
+  // accumulate reports
+  if (any_accumulate) {
+     profile("accumulate") {
+       reports = accumulate_reports(reports, accumulate);
```
Comment (seabbs) [inline inst/stan/estimate_infections.stan:145]: What is the interaction with truncation here? Before accumulation happened after truncation and now it happens before. I think it could be either  but perhaps after is slightly more common?

File: inst/stan/functions/observation_model.stan [hunk 8]
```diff
@@ -102,52 +102,54 @@ void truncation_lp(array[] real truncation_mean, array[] real truncation_sd,
...
+ * Accumulate reports according to a binary flag at each time point
+ *
+ * This function accumulates reports according to a binary flag at each time point.
+ *
+ * @param reports Vector of expected reports.
+ * @param accumulate Array of integers indicating, for each time point, whether to accumulate or not.
+ *
+ * @return A vector of accumulated reports.
+ */
+vector accumulate_reports(vector reports, array[] int accumulate) {
```
Comment (seabbs) [inline inst/stan/functions/observation_model.stan:144]: This is incredibly pedantic but I think this function should go above the truncation one match in model usage

**Review by seabbs — CHANGES_REQUESTED** [review]
This is really nice. I have a few small comments. At a high level I think we are missing examples of this in the wider docs that would be nice to add but could be done in another PR. However, as we are missing these I think we need to be careful here to make sure this does actually work as we are the only checks (i.e no correctness unit tests or figure to look at that I am aware of)

File: NEWS.md [hunk 9]
```diff
@@ -1,5 +1,20 @@
 # EpiNow2 (development version)
 
+## Model changes
+
+- The models now supports more complex accumulation patterns, as well as mixtures of accumulation and missingness via the new `fill_missing()` function and a logical `accumulate` column that can be supplied with the data. By @sbfnk in # and reviewed by @.
```
[the remark this answers is not in the document]
Comment (seabbs) [reply NEWS.md:5]: I agree but we do need to be careful as we don't support some aggregation schemes that are not accumulation (i.e non-consecutive aggregation)

File: R/create.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (seabbs) [reply R/create.R:539]: wait can't we just do any directly?

File: R/estimate_infections.R [hunk 10]
```diff
@@ -17,8 +17,13 @@
 #' for an example of using `estimate_infections` within the `epinow` wrapper to
 #' estimate Rt for Covid-19 in a country from the ECDC data source.
 #'
-#' @param data A `<data.frame>` of confirmed cases (confirm) by date
-#' (date). `confirm` must be numeric and `date` must be in date format.
+#' @param data A `<data.frame>` of confirmed cases (confirm) by date (date).
+#' `confirm` must be numeric and `date` must be in date format. Optionally
+#' this can also have a logical `accumulate` column which indicates whether
+#' data should be added to the next data point. This is useful when modelling
```
Comment (jamesmbaazam) [inline R/estimate_infections.R:23]: Can we reword this to something simpler for example, 
```suggestion
#' modelled values should correspond to the same reporting pattern as the observed data. This is useful when modelling
```

Comment (jamesmbaazam) [inline R/estimate_infections.R:23]:
```suggestion
#' `confirm` must be numeric and `date` must be in date format. Optionally
#' this can also have a logical `accumulate` column which indicates which time points
#'  should be aggregated/accumulated. This is useful when modelling
```

File: R/estimate_infections.R [hunk 11]
```diff
@@ -176,10 +181,11 @@ estimate_infections <- function(data,
     data = dirty_reported_cases,
     cols_to_check = c("date", "confirm")
   )
+  # Fill misisng dates
```
Comment (jamesmbaazam) [inline R/estimate_infections.R:184]:
```suggestion
  # Fill missing dates
```

File: NEWS.md [hunk 12]
```diff
@@ -1,5 +1,20 @@
...
+
+  ```r
+  # Deprecated
+  data |>
+    estimate_infections(obs_opts(na = "accumulate"))
+
+  # Recommended workflow e.g. for weekly incidence data
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply NEWS.md:12]: Not in this case (as `na == "missing"` by default) - will try to clarify in the docs as you suggest below.

File: R/create.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/create.R:539]:
```suggestion
    any_accumulate = as.integer(any(accumulate)),
```

File: R/estimate_infections.R (hunk 10 above)
Comment (sbfnk) [reply R/estimate_infections.R:23]: will rephrase

File: NEWS.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (sbfnk) [reply NEWS.md:5]: Sounds like a good idea to mention here if we think it's something people search for but in other places we should decide on one term (and make sure we explain it well / better) as otherwise people might just be confused about whether they mean the same or not.

[the remark this answers is not in the document]
Comment (sbfnk) [reply NEWS.md:5]:
```suggestion
- The models now supports more complex patterns of aggregating reported cases by accumulating them over multiple time points, as well as mixtures of accumulation and missingness via the new `fill_missing()` function and a logical `accumulate` column that can be supplied with the data. By @sbfnk in #851 and reviewed by @seabbs and @jamesmbaazam..
```

File: R/preprocessing.R (hunk 2 above)
Comment (sbfnk) [reply R/preprocessing.R:9]: ok, will try to rephrase / clarify in the function doc.

> I think the thing that confused me most was `missing = "missing"`. Perhaps it should be `missing = "na"`?

Hmm yes that does look weird - the problem is that you could have a pattern like `missing = "missing", na = "accumulate"` which I think would be even more confusing if then we had `missing = "na"`.

I'd suggest to:
- replace the choices with `c("ignore", "accumulate", "zero")` (with ignoring meaning ignore in the likelihood)
- rename the arguments from `missing` and `na` to `missing_dates` and `na_obs`, respectively

File: R/preprocessing.R (hunk 6 above)
Comment (sbfnk) [reply R/preprocessing.R:81]: I hadn't considered this. We could add a `by` argument, scoringutils-style?

File: inst/stan/functions/observation_model.stan [hunk 13]
```diff
@@ -102,52 +102,54 @@ void truncation_lp(array[] real truncation_mean, array[] real truncation_sd,
...
+  int ot_h = num_elements(reports); // number of reporting time points modelled
+  vector[ot_h] accumulated_reports = reports;
+  for (i in 1:(ot_h - 1)) {
+    if (accumulate[i]) { // first observation gets ignored when accumulating
+      accumulated_reports[i + 1] += accumulated_reports[i];
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply inst/stan/functions/observation_model.stan:149]: I think this works as expected (because `accumulated_reports` is set to `reports` earlier:

``` r
nreports <- 9

reports <- sample(1:100, nreports)
accumulate <- rep(c(TRUE, TRUE, FALSE), times = nreports / 3)

accumulated_reports <- reports
for (i in 1:(length(reports) - 1)) {
  if (accumulate[i]) {
    accumulated_reports[i + 1] <- accumulated_reports[i + 1] +
      accumulated_reports[i]
  }
}

data.frame(
  accumulate = accumulate,
  reports = reports,
  accumulated_reports = accumulated_reports
)
#>   accumulate reports accumulated_reports
#> 1       TRUE      24                  24
#> 2       TRUE       2                  26
#> 3      FALSE      51                  77
#> 4       TRUE      82                  82
#> 5       TRUE      35                 117
#> 6      FALSE      21                 138
#> 7       TRUE      19                  19
#> 8       TRUE      39                  58
#> 9      FALSE       1                  59
```

<sup>Created on 2024-11-12 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by sbfnk** [comment]
> At a high level I think we are missing examples of this in the wider docs that would be nice to add but could be done in another PR. 

I agree, e.g. #772.

> However, as we are missing these I think we need to be careful here to make sure this does actually work as we are the only checks (i.e no correctness unit tests or figure to look at that I am aware of)

Yes, good point. I can add a snapshot test (which will obviously only help in the future) and I think @jamesmbaazam is looking at testing this in action.

File: inst/stan/functions/observation_model.stan (hunk 13 above)
[the remark this answers is not in the document]
Comment (sbfnk) [reply inst/stan/functions/observation_model.stan:149]: > Is there a reason to code it like this? 

None apart from my inability to conceive of anything more elegant (noting that #548 remains to be addressed in a future PR). What would you suggest?

> there is nothing stopping you having multiple days in the likelihood as they accumulate

You mean fitting to the cumulative data? Of course it's a bad idea but I wasn't considering actively stopping people from doing so. Perhaps we should. I was thinking of accumulation and use in the likelihood as orthogonal but we could also only evaluate the likelihood if `accumulate == FALSE` AND the data is non-NA, which would be the safer option.

[the remark this answers is not in the document]
Comment (seabbs) [reply inst/stan/functions/observation_model.stan:149]: > You mean fitting to the cumulative data? 

Yes. 

> we could also only evaluate the likelihood if accumulate == FALSE AND the data is non-NA, which would be the safer option.

I thought this was the choice we made in the design doc but I don't feel that strongly because as you may maybe people might have a reason to fit to cumulative data

File: NEWS.md [hunk 14]
```diff
@@ -1,5 +1,20 @@
 # EpiNow2 (development version)
 
+## Model changes
+
+- The models now supports more complex patterns of aggregating reported cases by accumulating them over multiple time points, as well as mixtures of accumulation and missingness via the new `fill_missing()` function and a logical `accumulate` column that can be supplied with the data. By @sbfnk in #851 and reviewed by @seabbs and @jamesmbaazam..
```
Comment (jamesmbaazam) [inline NEWS.md:5]:
```suggestion
- The models now support more complex patterns of aggregating reported cases by accumulating them over multiple time points, as well as mixtures of accumulation and missingness via the new `fill_missing()` function and a logical `accumulate` column that can be supplied with the data. By @sbfnk in #851 and reviewed by @seabbs and @jamesmbaazam..
```

File: R/preprocessing.R [hunk 15]
```diff
@@ -0,0 +1,152 @@
+##' Fills missing data in a data set to prepare it for use within the package
```
Comment (jamesmbaazam) [inline R/preprocessing.R:1]:
```suggestion
##' Fill missing data in a data set to prepare it for use within the package
```

File: R/preprocessing.R (hunk 2 above)
Comment (jamesmbaazam) [reply R/preprocessing.R:9]: > rename the arguments from missing and na to missing_dates and na_obs, respectively

I was coming to suggest this.

File: R/preprocessing.R [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+##'
+##' @param data Data frame with a `date` column. The other columns depend on the
+##'   model that the data are to be used, e.g. [estimate_infections()] or
+##'   [estimate_secondary()]. See the documentation there for the expected
+##'   format.
+##' @param missing Character. Options are "missing" (the default), "accumulate"
+##'   and "zero". This determines how missing dates in the data are interpreted.
+##'   If set to "missing", any missing dates in the observation data will be
+##'   interpreted as missing and skipped in the likelihood. If set to
+##'   "accumulate", modelled observations will be accumulated and added to the
```
Comment (jamesmbaazam) [inline R/preprocessing.R:13]:
```suggestion
##'   "accumulate", the date will be modelled but accumulated and added to the
```

File: R/preprocessing.R [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+##'   next non-missing data point. This can be used to model incidence data that
+##'   is reported at less than daily intervals. If set to "accumulate", the
```
Comment (jamesmbaazam) [inline R/preprocessing.R:15]:
```suggestion
##'   next non-missing data point. This can be used to model incidence data that
##'   is reported at irregular/inconsistent intervals. If set to "accumulate", the
```

File: R/preprocessing.R [hunk 18]
```diff
@@ -0,0 +1,152 @@
+##' Fills missing data in a data set to prepare it for use within the package
+##'
+##' @description `r lifecycle::badge("experimental")`
+##'
+##' @param data Data frame with a `date` column. The other columns depend on the
+##'   model that the data are to be used, e.g. [estimate_infections()] or
+##'   [estimate_secondary()]. See the documentation there for the expected
+##'   format.
+##' @param missing Character. Options are "missing" (the default), "accumulate"
+##'   and "zero". This determines how missing dates in the data are interpreted.
```
Comment (jamesmbaazam) [inline R/preprocessing.R:10]:
```suggestion
##'   and "zero". This determines how implicit missingness is treated, i.e., how missing dates in the data are interpreted.
```

File: R/preprocessing.R [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+##'   If set to "missing", any missing dates in the observation data will be
+##'   interpreted as missing and skipped in the likelihood. If set to
+##'   "accumulate", modelled observations will be accumulated and added to the
+##'   next non-missing data point. This can be used to model incidence data that
+##'   is reported at less than daily intervals. If set to "accumulate", the
+##'   first data point is not included in the likelihood (unless `initial` is
+##'   set to a non-zero value) but used only to reset modelled observations to
+##'   zero. If "zero" then all observations on missing dates will be assumed to
+##'   be of value 0.
+##' @param na Character. How to process dates that have NA values. The options
```
Comment (jamesmbaazam) [inline R/preprocessing.R:20]:
```suggestion
##' @param na Character. This determines how explicit missing values are treated, i.e., dates that have `NA` observations. The options
```

File: R/preprocessing.R [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+##'   available are the same ones as for the `missing` argument.
+##' @param initial_accumulate Integer. The number of initial dates to accumulate
+##'   if `missing` or a column name is set to `"accumulate"`. This number of
+##'   dates is added to the beginning of the data set to be accumulated onto the
+##'   first data point. This is useful, for example, for modelling weekly
```
Comment (jamesmbaazam) [inline R/preprocessing.R:25]:
```suggestion
##' @param initial_accumulate Integer. Only used if `missing = "accumulate"` or the input data has a column named `"accumulate"`. This is the number of
##'   dates to model and accumulate to the first date in the input data. This is useful, for example, for modelling weekly
```

File: R/preprocessing.R [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+##'   incidence data, in which case this should be set to 7. If accumulating and
+##'   the first data point is one to accumulate and this is not set, then that
```
Comment (jamesmbaazam) [inline R/preprocessing.R:27]:
```suggestion
##'   the first data point is one to accumulate and `initial_accumulate` is not set, then that
```

File: NEWS.md (hunk 14 above)
Comment (seabbs) [reply NEWS.md:5]: as implemented it is literally accumulation and you can have multiple accumulate elements in the likelihood which would not be the same as aggregation so if we stick with that I think we need to stick with the accumulate naming

File: R/preprocessing.R (hunk 4 above)
Comment (seabbs) [reply R/preprocessing.R:29]: I think the action I would like here is to give it the confirm default as suggested

File: R/preprocessing.R [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+                         obs_column) {
+  assert_data_frame(data)
+  assert_names(names(data), must.include = "date", disjunct.from = "accumulate")
+  if (!missing(obs_column)) {
+    assert_names(names(data), must.include = obs_column)
+  } else {
+    obs_column <- setdiff(colnames(data), "date")
+  }
+  assert_date(data$date, any.missing = FALSE)
+
```
Comment (jamesmbaazam) [inline R/preprocessing.R:64]: assert_character(missing)
assert_character(na)
assert_character(obs_column)
assert_integer(initial_accumulate, lower = 0)

Comment (jamesmbaazam) [inline R/preprocessing.R:64]: Need to make sure `initial_accumulate` is never below 0.
```suggestion
assert_character(missing)
assert_character(na)
assert_character(obs_column)
assert_integer(initial_accumulate, lower = 0)
```

File: R/preprocessing.R [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+  data[, accumulate := FALSE]
+  for (col in obs_column) {
+    if (missing == "accumulate") {
+      data[is.na(..present), accumulate := TRUE]
+    } else if (missing == "zero") {
+      data[is.na(..present), paste(col) := 0]
+    }
+  }
+  if (missing(initial_accumulate) &&
+      !data$accumulate[1] &&
```
Comment (jamesmbaazam) [inline R/preprocessing.R:92]:
```suggestion
      isFALSE(data$accumulate[1]) &&
```

File: R/preprocessing.R [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+##' ## limit to dates once a week
+##' cases <- cases[seq(7, nrow(cases), 7)]
+##' ## set the second observation to missing
+##' cases[2, confirm := NA]
+##' ## fill missing data
+##' fill_missing(cases, missing = "accumulate", initial_accumulate = 7)
+fill_missing <- function(data,
+                         missing = c("missing", "accumulate", "zero"),
+                         na = c("missing", "accumulate", "zero"),
+                         initial_accumulate,
```
Comment (jamesmbaazam) [inline R/preprocessing.R:54]: I'm a little confused. I set `initial_accumulate = 1` and expected 2020-02-28  to be set to TRUE and an observation added at 2020-02-27. Is this expected?

```r
library(data.table)
cases <- data.table::copy(example_confirmed)
## calculate weekly sum
cases[, confirm := data.table::frollsum(confirm, 7)]
## limit to dates once a week
cases <- cases[seq(7, nrow(cases), 7)]
## set the second observation to missing
cases[2, confirm := NA]
## fill missing data
fill_missing(cases, missing = "accumulate", initial_accumulate = 1) |>
  head(14)

Key: <date>
          date confirm accumulate
        <Date>   <num>     <lgcl>
 1: 2020-02-28     647      FALSE
 2: 2020-02-29      NA       TRUE
 3: 2020-03-01      NA       TRUE
  
```

Comment (jamesmbaazam) [reply R/preprocessing.R:54]: Also, what should be the expected behavior if `intial_accumulate` is set to 0? Should it be the same as when it is not set?

Comment (sbfnk) [reply R/preprocessing.R:54]: The idea is that the default is to accumulate one observation (`initial_accumulate = 1`). 

> 2020-02-28 to be set to TRUE and an observation added at 2020-02-27. 

This to me sounds like two observations are being accumulated (`initial_accumulate = 2`)

> Also, what should be the expected behavior if initial_accumulate is set to 0? Should it be the same as when it is not set?

We should probably make it have a minimum of 1.

Happy to hear of alternative/clearer suggestions for wording this.

Comment (jamesmbaazam) [reply R/preprocessing.R:54]: > We should probably make it have a minimum of 1.

Agreed. That can be added in the assertion on `initial_accumulate` I added below.

File: R/preprocessing.R [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+##' cases <- data.table::copy(example_confirmed)
+##' ## calculate weekly sum
+##' cases[, confirm := data.table::frollsum(confirm, 7)]
+##' ## limit to dates once a week
+##' cases <- cases[seq(7, nrow(cases), 7)]
+##' ## set the second observation to missing
+##' cases[2, confirm := NA]
+##' ## fill missing data
+##' fill_missing(cases, missing = "accumulate", initial_accumulate = 7)
+fill_missing <- function(data,
```
Comment (jamesmbaazam) [inline R/preprocessing.R:51]: This function errors if the data already has an `accumulate` column (from the assertion for column names). 

Should it rather throw a message and do nothing if it finds an `accumulate` column?

```r
library(data.table)
cases <- data.table::copy(example_confirmed)
## calculate weekly sum
cases[, confirm := data.table::frollsum(confirm, 7)]
## limit to dates once a week
cases <- cases[seq(7, nrow(cases), 7)]
## set the second observation to missing
cases[2, confirm := NA][, accumulate:=TRUE]
## fill missing data
fill_missing(cases, missing = "accumulate", initial_accumulate = 1, obs_column = "confirm")

Error in fill_missing(cases, missing = "accumulate", initial_accumulate = 1,  : 
  Assertion on 'names(data)' failed: Names must be disjunct from {'accumulate'}, but has elements {'accumulate'}.
```

File: NEWS.md (hunk 14 above)
Comment (sbfnk) [reply NEWS.md:5]:
```suggestion
- The models now support more complex patterns of aggregating reported cases by accumulating them over multiple time points, as well as mixtures of accumulation and missingness via the new `fill_missing()` function and a logical `accumulate` column that can be supplied with the data. By @sbfnk in #851 and reviewed by @seabbs and @jamesmbaazam..
```

File: R/preprocessing.R (hunk 25 above)
Comment (sbfnk) [reply R/preprocessing.R:59]: What would be the rationale for this? I don't really see when we'd want to fail silently if someone calls `fill_missing()` and we're not doing anything because an `accumulate` column already exists.

File: R/preprocessing.R (hunk 24 above)
Comment (sbfnk) [reply R/preprocessing.R:62]: I've updated the documentation in the attempt to make it clearer

File: R/preprocessing.R (hunk 17 above)
Comment (sbfnk) [reply R/preprocessing.R]: I've tried to rephrase slightly differently in an attempt to make it clearer.

File: R/preprocessing.R (hunk 19 above)
Comment (sbfnk) [reply R/preprocessing.R]: Have tried to rephrase.

File: R/preprocessing.R (hunk 18 above)
Comment (sbfnk) [reply R/preprocessing.R]: I think if doing this we'd have to define implicit vs. explicit missingness somewhere - do you think it would make it clearer? I have tried to rephrase without introducing new definitions but please comment if you disagree with the approach.

File: R/preprocessing.R (hunk 20 above)
Comment (sbfnk) [reply R/preprocessing.R]: Have tried to rephrase

**Comment by sbfnk** [comment]
Thanks for the comments/suggestions - I've updated the argument names as suggested and did a pass through the docs in an attempt to make them clearer.

File: R/preprocessing.R (hunk 25 above)
Comment (sbfnk) [reply R/preprocessing.R:59]: Alternatively perhaps we could have a more informative error message?

Comment (sbfnk) [reply R/preprocessing.R:59]: Alternatively perhaps we could have a more informative error message?

Comment (seabbs) [reply R/preprocessing.R:59]: I am pretty happy with it failing as `fill_missing` is optional. We could make it more informative but tbh I am happy

File: inst/stan/estimate_infections.stan [hunk 26]
```diff
@@ -156,6 +156,13 @@ transformed parameters {
   } else {
     obs_reports = reports[1:ot];
   }
+
+  // accumulate reports
+  if (any_accumulate) {
```
Comment (seabbs) [inline inst/stan/estimate_infections.stan:161]: Just checking the difference between estimate_infections and truncation is intentional (i.e one accumulates after truncation and the other before)

**Review by seabbs — DISMISSED** [review]
This looks good to me and happy for it to be merged. I have a open comment with a last check about the interaction between accumulation and truncation (which I think I already raised and so is likely easy to dismiss)

**Comment by seabbs** [comment]
CI failures are unrelated to this PR (trying and failing to download R) so I am happy to ignore them

File: R/preprocessing.R (hunk 25 above)
Comment (jamesmbaazam) [reply R/preprocessing.R:59]: > Alternatively perhaps we could have a more informative error message?

Yes, something like "data already has an "accumulate" column" is more friendly.

**Review by seabbs — APPROVED** [review]

File: R/preprocessing.R (hunk 25 above)
Comment (sbfnk) [reply R/preprocessing.R:66]: I've decided against this as really this applies to all the `{checkmate}` checks and we don't want to change them all. I added some more clarifying documentation instead on the topic.

**Comment by sbfnk** [comment]
Thanks all for the comments/reviews - merging now as failed checks are unrelated.
